iree/compiler/Dialect/HAL/IR/HALOps.td

```diff
@@ -2032,6 +2026,115 @@ def HAL_ExecutableLayoutLookupOp : HAL_PureOp<"executable_layout.lookup", [
 // iree::hal::Semaphore
 //===----------------------------------------------------------------------===//
 
-// TODO(benvanik): semaphore query/set.
+def HAL_SemaphoreCreateOp : HAL_Op<"semaphore.create", [
+    DeclareOpInterfaceMethods<OpAsmOpInterface>,
+  ]> {
+  let summary = [{semaphore allocation operation}];
+  let description = [{
+    Returns a semaphore from the device pool with the given initial value.
+  }];
+
+  let arguments = (ins
+    HAL_Device:$device,
+    HAL_TimelineValue:$initial_value
+  );
+  let results = (outs
+    HAL_Semaphore:$result
+  );
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<[{
+      Builder *builder, OperationState &state, Value device,
+      int32_t initial_value
+    }]>,
+  ];
+
+  let assemblyFormat = [{
+    $device `,` `initial_value` `=` $initial_value
+    attr-dict-with-keyword `:` type($result)
+  }];
+}
+
+def HAL_SemaphoreQueryOp : HAL_Op<"semaphore.query"> {
+  let summary = [{semaphore payload value query}];
+  let description = [{
+    Queries the current payload and returns a tuple of `(status, value)`.
+    As the payload is monotonically increasing it is guaranteed that
+    the value is at least equal to the previous result of a
+    `hal.semaphore.signal` call and coherent with any waits for a
+    specified value via `hal.semaphore.await`.
+  }];
+
+  let arguments = (ins
+    HAL_Semaphore:$semaphore
+  );
+  let results = (outs
+    IREE_Status:$status,
+    HAL_TimelineValue:$value
+  );
+
+  let assemblyFormat = [{
+    $semaphore attr-dict-with-keyword `:` type($status) `,` type($value)
+  }];
+}
+
+def HAL_SemaphoreSignalOp : HAL_Op<"semaphore.signal"> {
+  let summary = [{semaphore payload value signal operation}];
+  let description = [{
+    Signals the semaphore to the given payload value.
+    The call is ignored if the current payload value exceeds `new_value`.
```

Comment:
On our side yes because the values must increase to be signaled signaling to an earlier value should be a no-op. I don't think Vulkan will like that though, so we may need to sanitize things by keeping track of the last value we signaled to and doing the no-op ourselves. It may be a thing we only enable when we are running untrusted models though.